File: getting_started/grpc/grpc.js
import fs from 'fs';
import path from 'path';
import grpc from '@grpc/grpc-js';
import jwt from 'jsonwebtoken';
import { RogueDBClient } from './getting_started/roguedb_grpc_pb.js'; 
import rogue_pb from './getting_started/roguedb_pb.js';
import google_protobuf_any_pb from 'google-protobuf/google/protobuf/any_pb.js';

const { Test, Insert, Search, Basic, LogicalOperator, ComparisonOperator, Subscribe } = rogue_pb;

async function createJwt(serviceAccountPath, expireMinutes = 60) {
  const raw = await fs.readFile(serviceAccountPath, 'utf8');
  const keyData = JSON.parse(raw);

  const now = Math.floor(Date.now() / 1000);
  const payload = {
    iat: now,
    exp: now + expireMinutes * 60,
    iss: keyData.client_email,
    sub: keyData.client_email,
    aud: `${keyData.client_email.split('@')[0]}.roguedb.dev`,
  };

  // keyid placed in header automatically if provided via options.keyid
  return jwt.sign(payload, keyData.private_key, {
    algorithm: 'RS256',
    keyid: keyData.private_key_id,
    header: { typ: 'JWT' },
  });
}

function detectFiles(dir, fileList = []) {
    const files = fs.readdirSync(dir);
    files.forEach(file => {
        const filePath = path.join(dir, file);
        if (fs.statSync(filePath).isDirectory()) {
            detectFiles(filePath, fileList);
        } else if (path.extname(file) === '.proto') {
            fileList.push(filePath);
        }
    });
    return fileList;
}

async function main() {
    // See purchase confirmation emails for details and service_account.json.
    const API_KEY = "API_KEY";
    const URL = "DATABASE_URL";
    const ENCODED_JWT = createJwt();

    const client = new RogueDBClient(
        `${URL}:443`,
        grpc.credentials.createSsl()
    );

    const metadata = new grpc.Metadata();
    metadata.add('Authorization', `Bearer ${ENCODED_JWT}`);

    ////////////////////////////////////////////////////////
    ///////  Insert, Update, and Remove API Example  ///////
    ////////////////////////////////////////////////////////

    const test = new Test();
    test.setAttribute1(10);

    const anyMessage = new google_protobuf_any_pb.Any();
    anyMessage.pack(test.serializeBinary(), 'rogue.services.Test');

    const request = new Insert();
    // const request = new Update();
    // const request = new Remove();
    request.setApiKey(API_KEY);
    request.addMessages(anyMessage);

    // No response is given for Insert, Update, and Remove
    // Any errors get reported in status.
    const stream = client.insert(metadata, (error) => { // Insert API
    // const stream = client.update(metadata, (error) => { // Update API
    // const stream = client.remove(metadata, (error) => { // Remove API
        if (error) console.error("Insert Status Error:", error);
    });
    stream.write(request);
    stream.end();

    //////////////////////////////////////
    ////////  Search API Example  ////////
    //////////////////////////////////////

    const search = new Search();
    search.setApiKey(API_KEY);
    
    // Example of a basic index query.
    // For Test, attribute1, attribute2, and attribute3 form the index.
    // Search Query:
    // Test.attribute1 >= 1 and Test.attribute2 >= 1 and Test.attribute3 >= true
    // AND
    // Test.attribute1 <= 10 and Test.attribute2 <= 10 and Test.attribute3 <= true
    const expression = new Basic();
    expression.setLogicalOperator(LogicalOperator.AND);
    expression.addComparisons(ComparisonOperator.GREATER_EQUAL);
    expression.addComparisons(ComparisonOperator.LESSER_EQUAL);
    
    const testGte = new Test();
    testGte.setAttribute1(1);
    testGte.setAttribute2(1);
    testGte.setAttribute3(true);
    
    const gteAny = new google_protobuf_any_pb.Any();
    gteAny.pack(testGte.serializeBinary(), 'rogue.services.Test');
    expression.addOperands(gteAny);

    const testLte = new Test();
    testLte.setAttribute1(10);
    testLte.setAttribute2(10);
    testLte.setAttribute3(true);
    
    const lteAny = new google_protobuf_any_pb.Any();
    lteAny.pack(testLte.serializeBinary(), 'rogue.services.Test');
    expression.addOperands(lteAny);

    search.addQueries().setBasic(expression);

    const searchStream = client.search(metadata);
    searchStream.on('data', (response) => {
        // Queries are zero-indexed. Partial results get sent
        // and mapped to that index.
        const results = response.getResultsList()[0].getMessagesList();
        results.forEach(resAny => {
            const resultTest = resAny.unpack(Test.deserializeBinary, 'rogue.services.Test');
            console.log("Unpacked Attribute1:", resultTest.getAttribute1());
        });
    });
    searchStream.write(search);
    searchStream.end(); // Signal all queries sent. Otherwise, blocks.

    // Example of a basic non-indexed query.
    // Search Query: 
    // Test.attribute1 >= 1 && Test.attribute2 <= 10
    const expression2 = new Basic();
    expression2.setLogicalOperator(LogicalOperator.AND);
    expression2.addComparisons(ComparisonOperator.GREATER_EQUAL);
    expression2.addComparisons(ComparisonOperator.LESSER_EQUAL);
    expression2.addFields(1); // Corresponds to attribute1 field id in Test
    expression2.addFields(2); // Corresponds to attribute2 field id in Test
    
    const testGte2 = new Test();
    testGte2.setAttribute1(1);
    
    const gteAny2 = new google_protobuf_any_pb.Any();
    gteAny2.pack(testGte2.serializeBinary(), 'rogue.services.Test');
    expression2.addOperands(gteAny2);

    const testLte2 = new Test();
    testLte2.setAttribute1(10);
    
    const lteAny2 = new google_protobuf_any_pb.Any();
    lteAny2.pack(testLte2.serializeBinary(), 'rogue.services.Test');
    expression2.addOperands(lteAny2);

    search.addQueries().setBasic(expression2);

    ///////////////////////////////////
    //// Schema Change API Example ////
    ///////////////////////////////////

    const subscribe = new Subscribe();
    subscribe.setApiKey(API_KEY);
    detectFiles('/path/to/protos').forEach(file => {
        // All proto files should be sent in a list of
        // their contents. No modifications required.
        subscribe.addSchemas(fs.readFileSync(file, 'utf8'));
    });

    // Any schemas excluded will have associated data deleted.
    // Schema change failure results in no changes applied.
    client.subscribe(subscribe, metadata, (err, response) => {
        if (!err) console.log("Schema updated successfully");
    });
}

main();

File: getting_started/grpc/grpc.java
import com.auth0.jwt.JWT;
import com.auth0.jwt.algorithms.Algorithm;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;

import java.nio.file.Files;
import java.nio.file.Paths;
import java.nio.file.Path;
import java.util.List;
import java.util.Base64;
import java.security.spec.PKCS8EncodedKeySpec;
import java.security.interfaces.RSAPrivateKey;
import java.security.KeyFactory;
import java.time.Instant;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.TimeUnit;
import java.util.stream.Collectors;

import com.google.protobuf.Any;
import io.grpc.ManagedChannel;
import io.grpc.ManagedChannelBuilder;
import io.grpc.Metadata;
import io.grpc.stub.MetadataUtils;
import io.grpc.stub.StreamObserver;

import rogue.services.RogueDBGrpc;
import rogue.services.Roguedb.*;

public class grpc {
    private static final ObjectMapper mapper = new ObjectMapper();
    
    public static String createJwt(String serviceAccountPath) throws Exception {
        byte[] jsonData = Files.readAllBytes(Paths.get(serviceAccountPath));
        var keyData = mapper.readTree(jsonData);

        String privateKeyContent = keyData.get("private_key").asText()
                .replace("-----BEGIN PRIVATE KEY-----", "")
                .replace("-----END PRIVATE KEY-----", "")
                .replaceAll("\\s", "");

        byte[] encoded = Base64.getDecoder().decode(privateKeyContent);
        PKCS8EncodedKeySpec keySpec = new PKCS8EncodedKeySpec(encoded);
        RSAPrivateKey privateKey = (RSAPrivateKey) KeyFactory.getInstance("RSA").generatePrivate(keySpec);

        Algorithm algorithm = Algorithm.RSA256(null, privateKey);
        String email = keyData.get("client_email").asText();
        String aud = email.split("@")[0] + ".roguedb.dev";

        return JWT.create()
                .withKeyId(keyData.get("private_key_id").asText())
                .withIssuer(email)
                .withSubject(email)
                .withAudience(aud)
                .withIssuedAt(Instant.now())
                .withExpiresAt(Instant.now().plusSeconds(3600))
                .sign(algorithm);
    }

    public static List<String> detectFiles(Path dir) throws Exception {
        try (var stream = Files.walk(dir)) {
            return stream
                .filter(p -> p.toString().endsWith(".proto"))
                .map(p -> {
                    try { return Files.readString(p); } 
                    catch (Exception e) { return ""; }
                })
                .collect(Collectors.toList());
        }
    }

    public static void main(String[] args) throws Exception {
        // See purchase confirmation emails for details and service_account.json.
	    final String API_KEY = "YOUR_API_KEY";
        final String URL = "c-[YOUR_IDENTIFIER].roguedb.dev";
        final String JWT_TOKEN = createJwt("/path/to/service_account.json");

        ManagedChannel channel = ManagedChannelBuilder.forAddress(URL, 443)
            .useTransportSecurity()
            .build();

        // Setup Headers
        Metadata headers = new Metadata();
        headers.put(Metadata.Key.of("Authorization", Metadata.ASCII_STRING_MARSHALLER), "Bearer " + JWT_TOKEN);
        
        // Use the Async Stub
        RogueDBGrpc.RogueDBStub asyncStub = RogueDBGrpc.newStub(channel)
            .withInterceptors(MetadataUtils.newAttachHeadersInterceptor(headers));

        ////////////////////////////////////////////////////////
        ///////  Insert, Update, and Remove API Example  ///////
        ////////////////////////////////////////////////////////
        
        final CountDownLatch finishLatch = new CountDownLatch(1);
        // Insert API
        StreamObserver<Insert> stream = asyncStub.insert(new StreamObserver<Response>() {
        // StreamObserver<Update> stream = asyncStub.update(new StreamObserver<Response>() { // Update API
        // StreamObserver<Remove> stream = asyncStub.remove(new StreamObserver<Response>() { // Remove API
            
            // No response is given for Insert, Update, and Remove
            // Any errors get reported in status.
            @Override public void onNext(Response value) { /* No-op for Insert */ }
            @Override public void onError(Throwable t) { finishLatch.countDown(); }
            @Override public void onCompleted() { finishLatch.countDown(); }
        });

        stream.onNext(Insert.newBuilder()
        // stream.onNext(Update.newBuilder() // Update API
        // stream.onNext(Remove.newBuilder() // Remove API
            .setApiKey(API_KEY)
            // Insert, Update, and Remove are identical in use.
            .addMessages(Any.pack(Test.newBuilder().setAttribute1(10).build()))
            .build());
        stream.onCompleted(); // Signal all queries sent. Otherwise, blocks.
        finishLatch.await(1, TimeUnit.MINUTES);

        //////////////////////////////////////
        ////////  Search API Example  ////////
        //////////////////////////////////////
        
        final CountDownLatch searchFinishLatch = new CountDownLatch(1);
        StreamObserver<Search> searchStream = asyncStub.search(new StreamObserver<Response>() {
            @Override
            public void onNext(Response response) {
                // Queries are zero-indexed. Partial results get sent
                // and mapped to that index.
                response.getResultsMap().get(0L).getMessagesList().forEach(any -> {
                    try {
                        Test t = any.unpack(Test.class);
                    } catch (Exception e) { e.printStackTrace(); }
                });
            }
            @Override public void onError(Throwable t) { searchFinishLatch.countDown(); }
            @Override public void onCompleted() { searchFinishLatch.countDown(); }
        });

        // Example of a basic index query.
        // For Test, attribute1, attribute2, and attribute3 form the index.
        // Search Query:
        // Test.attribute1 >= 1 and Test.attribute2 >= 1 and Test.attribute3 >= true
        // AND
        // Test.attribute1 <= 10 and Test.attribute2 <= 10 and Test.attribute3 <= true
        searchStream.onNext(Search.newBuilder()
            .setApiKey(API_KEY)
            .addQueries(Expression.newBuilder()
                .setBasic(Basic.newBuilder()
                    .setLogicalOperator(LogicalOperator.AND)
                    .addComparisons(ComparisonOperator.GREATER_EQUAL)
                    .addOperands(Any.pack(Test.newBuilder()
                        .setAttribute1(1)
                        .setAttribute2(1)
                        .setAttribute3(true)
                        .build()))
                    .addComparisons(ComparisonOperator.LESSER_EQUAL)
                    .addOperands(Any.pack(Test.newBuilder()
                        .setAttribute1(10)
                        .setAttribute2(10)
                        .setAttribute3(true)
                        .build()))))
            .build());
        searchStream.onCompleted(); // Signal all queries sent. Otherwise, blocks.
        searchFinishLatch.await(1, TimeUnit.MINUTES);

        // Example of a basic non-indexed query.
        // Search Query: 
        // Test.attribute1 >= 1 && Test.attribute2 <= 10
        searchStream.onNext(Search.newBuilder()
            .setApiKey(API_KEY)
            .addQueries(Expression.newBuilder()
                .setBasic(Basic.newBuilder()
                    .setLogicalOperator(LogicalOperator.AND)
                    .addFields(1)
                    .addFields(2)
                    .addComparisons(ComparisonOperator.GREATER_EQUAL)
                    .addOperands(Any.pack(Test.newBuilder()
                        .setAttribute1(1)
                        .setAttribute2(1)
                        .setAttribute3(true)
                        .build()))
                    .addComparisons(ComparisonOperator.LESSER_EQUAL)
                    .addOperands(Any.pack(Test.newBuilder()
                        .setAttribute1(10)
                        .setAttribute2(10)
                        .setAttribute3(true)
                        .build()))))
            .build());

        ///////////////////////////////////
        //// Schema Change API Example ////
        ///////////////////////////////////
        
        var blockingStub = RogueDBGrpc.newBlockingStub(channel).withInterceptors(
            MetadataUtils.newAttachHeadersInterceptor(headers));

        Subscribe.Builder subscribeBuilder = Subscribe.newBuilder().setApiKey(API_KEY);
        // All proto files should be sent in a list of
        // their contents. No modifications required.
        for(String file : detectFiles(Paths.get("/path/to/protos"))) {
            subscribeBuilder.addSchemas(Files.readString(Paths.get(file)));
        }

        // Any schemas excluded will have associated data deleted.
        // Schema change failure results in no changes applied.
        Response subscribeResponse = blockingStub.subscribe(subscribeBuilder.build());
    }
}
